kin-ecosystem-sdk/src/main/java/com/kin/ecosystem/marketplace/presenter/MarketplacePresenter.java

```diff
@@ -39,98 +41,113 @@
 
-	private List<Offer> spendList = new ArrayList<>();
-	private List<Offer> earnList = new ArrayList<>();
+	private List<Offer> spendList;
+	private List<Offer> earnList;
 
-	private Observer<Offer> pendingOfferObserver;
-	private Observer<Order> completedOrderObserver;
-
-	private boolean isEarnListEmpty;
-	private boolean isSpendListEmpty;
+	private Observer<Order> orderObserver;
 
 
 	private final Gson gson;
 
-	public MarketplacePresenter(@NonNull final OfferDataSource offerRepository,
-		@NonNull final OrderDataSource orderRepository, @Nullable final BlockchainSource blockchainSource, @NonNull
-		EventLogger eventLogger) {
-		this.spendList = new ArrayList<>();
-		this.earnList = new ArrayList<>();
+	public MarketplacePresenter(@NonNull final IMarketplaceView view, @NonNull final OfferDataSource offerRepository,
+		@NonNull final OrderDataSource orderRepository, @Nullable final BlockchainSource blockchainSource,
+		@NonNull INavigator navigator, @NonNull EventLogger eventLogger) {
+		this.view = view;
 		this.offerRepository = offerRepository;
 		this.orderRepository = orderRepository;
 		this.blockchainSource = blockchainSource;
+		this.navigator = navigator;
 		this.eventLogger = eventLogger;
 		this.gson = new Gson();
+
+		this.view.attachPresenter(this);
 	}
 
 	@Override
 	public void onAttach(IMarketplaceView view) {
 		super.onAttach(view);
 		getCachedOffers();
 		getOffers();
-		listenToPendingOffers();
 		listenToCompletedOrders();
 		eventLogger.send(MarketplacePageViewed.create());
 	}
 
 	private void getCachedOffers() {
-		OfferList cachedOfferList = offerRepository.getCachedOfferList();
-		setOfferList(cachedOfferList);
+		if(earnList == null && spendList == null) {
+			earnList = new ArrayList<>();
+			spendList = new ArrayList<>();
+			OfferList cachedOfferList = offerRepository.getCachedOfferList();
+			if (cachedOfferList != null && cachedOfferList.getOffers() != null) {
+				splitOffersByType(cachedOfferList.getOffers(), this.earnList, this.spendList);
+			}
+		}
+		setOfferLists();
 	}
 
-	private void listenToCompletedOrders() {
-		completedOrderObserver = new Observer<Order>() {
-			@Override
-			public void onChanged(Order order) {
-				getOffers();
-			}
-		};
-		orderRepository.addCompletedOrderObserver(completedOrderObserver);
+	private void setOfferLists() {
+		if (this.view != null) {
+			this.view.setEarnList(earnList);
+			this.view.setSpendList(spendList);
+
+			setEarnEmptyViewState();
+			setSpendEmptyViewState();
+		}
 	}
 
-	private void listenToPendingOffers() {
-		pendingOfferObserver = new Observer<Offer>() {
+	private void listenToCompletedOrders() {
+		orderObserver = new Observer<Order>() {
 			@Override
-			public void onChanged(Offer offer) {
-				if (offer != null) {
-					removeOfferFromList(offer);
+			public void onChanged(Order order) {
+				switch (order.getStatus()) {
+					case PENDING:
+						removeOfferFromList(order.getOfferId(), order.getOfferType());
```

Comment:
this will now only be called after submit order request completed - (there can be lag in the UI) right?

Reply:
I won't call it lag, a slight delay.